src/data/simple_dmatrix.cc

```diff
@@ -47,28 +48,44 @@ class SimpleBatchIteratorImpl : public BatchIteratorImpl {
   }
 
  private:
-  SparsePage* page_{nullptr};
+  T* page_{nullptr};
 };
 
-BatchSet SimpleDMatrix::GetRowBatches() {
+template<typename T>
+BatchSet<T> SimpleDMatrix::GetSimpleBatches(PageType page_type) {
+  switch (page_type) {
+    case kCSR:
+      return GetRowBatches();
+    case kCSC:
+      return GetColumnBatches();
+    case kSortedCSC:
+      return GetSortedColumnBatches();
+    default:
+      LOG(FATAL) << "Unknown page type: " << page_type;
+      return BatchSet<T>(BatchIterator<T>(nullptr));
+  }
+}
+
+BatchSet<SparsePage> SimpleDMatrix::GetRowBatches() {
   auto cast = dynamic_cast<SimpleCSRSource*>(source_.get());
-  auto begin_iter = BatchIterator(new SimpleBatchIteratorImpl(&(cast->page_)));
-  return BatchSet(begin_iter);
+  auto begin_iter = BatchIterator<SparsePage>(
+      new SimpleBatchIteratorImpl<SparsePage>(&(cast->page_)));
+  return BatchSet<SparsePage>(begin_iter);
 }
 
-BatchSet SimpleDMatrix::GetColumnBatches() {
+BatchSet<SparsePage> SimpleDMatrix::GetColumnBatches() {
```

Comment:
multiple implementations of this have to exist for every page type - let's say if you need a compressed page type in the future.